Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.h

```diff
@@ -54,38 +84,53 @@ class WebXROpaqueFramebuffer {
     static std::unique_ptr<WebXROpaqueFramebuffer> create(PlatformXR::LayerHandle, WebGLRenderingContextBase&, Attributes&&, IntSize);
     ~WebXROpaqueFramebuffer();
 
+    bool supportsDynamicViewportScaling() const;
+
     PlatformXR::LayerHandle handle() const { return m_handle; }
-    const WebGLFramebuffer& framebuffer() const { return m_framebuffer.get(); }
-    GCGLint width() const { return m_framebufferSize.width(); }
-    GCGLint height() const { return m_framebufferSize.height(); }
+    const WebGLFramebuffer& framebuffer() const { return m_drawFramebuffer.get(); }
+    IntSize displayFramebufferSize() const;
+    IntSize drawFramebufferSize() const;
+    IntRect drawViewport(PlatformXR::Eye) const;
 
     void startFrame(const PlatformXR::FrameData::LayerData&);
     void endFrame();
 
 private:
     WebXROpaqueFramebuffer(PlatformXR::LayerHandle, Ref<WebGLFramebuffer>&&, WebGLRenderingContextBase&, Attributes&&, IntSize);
 
-    bool setupFramebuffer();
-    PlatformGLObject allocateRenderbufferStorage(GraphicsContextGL&, GCGLsizei, GCGLenum, IntSize);
-    PlatformGLObject allocateColorStorage(GraphicsContextGL&, GCGLsizei, IntSize);
-    PlatformGLObject allocateDepthStencilStorage(GraphicsContextGL&, GCGLsizei, IntSize);
-    void bindColorBuffer(GraphicsContextGL&, PlatformGLObject);
-    void bindDepthStencilBuffer(GraphicsContextGL&, PlatformGLObject);
+#if PLATFORM(COCOA)
+    bool setupFramebuffer(GraphicsContextGL&, const PlatformXR::FrameData::LayerSetupData&);
+#endif
```

Comment:
This code is really need for the non-Cocoa platforms too. With the countless refactoring, OpenXR is probably very broken. We should split the code to have WebXROpaqueFramebufferCocoa...

Reply:
`LayerSetupData` only exists for `PLATFORM(COCOA)`. Should it be added for all platforms?

Reply:
The frame buffer setup should occur in `WebXROpaqueFramebuffer::startFrame` for OpenXR